Build failure: python311Packages.xformers with CUDA

### Steps To Reproduce

Symptoms practically identical to #288621.

Steps to reproduce the behavior:
1. build `python311Packages.xformers` (e.g. on `nixpkgs-unstable`) when `config.cudaSupport` is enabled.

### Build log

Relevant section ([full log](https://github.com/user-attachments/files/15931190/log.txt)):

```
building 'xformers._C_flashattention' extension
creating build/temp.linux-x86_64-cpython-311
creating build/temp.linux-x86_64-cpython-311/third_party
creating build/temp.linux-x86_64-cpython-311/third_party/flash-attention
creating build/temp.linux-x86_64-cpython-311/third_party/flash-attention/csrc
creating build/temp.linux-x86_64-cpython-311/third_party/flash-attention/csrc/flash_attn
creating build/temp.linux-x86_64-cpython-311/third_party/flash-attention/csrc/flash_attn/src
g++ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -I/nix/store/gk0hrl9rngz3lfrnisql0h4xm65p036z-libxcrypt-4.4.36/include -fPIC -I/build/source/third_party/flash-attention/c>
third_party/flash-attention/csrc/flash_attn/flash_api.cpp: In function ‘void set_params_fprop(Flash_fwd_params&, size_t, size_t, size_t, size_t, size_t, size_t, size_t, si>
third_party/flash-attention/csrc/flash_attn/flash_api.cpp:48:11: warning: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘struct Flash_fwd_param>
   48 |     memset(&params, 0, sizeof(params));
      |     ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from third_party/flash-attention/csrc/flash_attn/flash_api.cpp:13:
/build/source/third_party/flash-attention/csrc/flash_attn/src/flash.h:51:8: note: ‘struct Flash_fwd_params’ declared here
   51 | struct Flash_fwd_params : public Qkv_params {
      |        ^~~~~~~~~~~~~~~~
/nix/store/fnabmiiyrfqqps870i19139zbvvc5g6j-cuda_nvcc-12.2.140/bin/nvcc -I/build/source/third_party/flash-attention/csrc/flash_attn -I/build/source/third_party/flash-atten>
In file included from /nix/store/avkdrmgnn4c0b4sqnlgbn0w464cavh33-cuda_cudart-12.2.140/include/cuda_runtime.h:82,
                 from <command-line>:
/nix/store/fnabmiiyrfqqps870i19139zbvvc5g6j-cuda_nvcc-12.2.140/include/crt/host_config.h:143:2: error: #error -- unsupported GNU version! gcc versions later than 12 are no>
  143 | #error -- unsupported GNU version! gcc versions later than 12 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version c>
      |  ^~~~~
In file included from /nix/store/avkdrmgnn4c0b4sqnlgbn0w464cavh33-cuda_cudart-12.2.140/include/cuda_runtime.h:82,
                 from <command-line>:
/nix/store/fnabmiiyrfqqps870i19139zbvvc5g6j-cuda_nvcc-12.2.140/include/crt/host_config.h:143:2: error: #error -- unsupported GNU version! gcc versions later than 12 are no>
  143 | #error -- unsupported GNU version! gcc versions later than 12 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version c>
      |  ^~~~~
error: command '/nix/store/fnabmiiyrfqqps870i19139zbvvc5g6j-cuda_nvcc-12.2.140/bin/nvcc' failed with exit code 255
```

### Additional context

### Notify maintainers

@happysalada

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/xformers/default.nix
+++ b/pkgs/development/python-modules/xformers/default.nix
@@ -1,5 +1,6 @@
 {
   lib,
+  stdenv,
   buildPythonPackage,
   pythonOlder,
   fetchFromGitHub,
@@ -52,11 +53,13 @@ buildPythonPackage {
       # noqa: C801
       __version__ = "${version}"
       EOF
-    ''
-    + lib.optionalString cudaSupport ''
-      export CUDA_HOME=${cudaPackages.cuda_nvcc}
-      export TORCH_CUDA_ARCH_LIST="${lib.concatStringsSep ";" cudaCapabilities}"
     '';
+
+  env = lib.attrsets.optionalAttrs cudaSupport {
+    TORCH_CUDA_ARCH_LIST = "${lib.concatStringsSep ";" torch.cudaCapabilities}";
+  };
+
+  stdenv = if cudaSupport then cudaPackages.backendStdenv else stdenv;
 
   buildInputs = lib.optionals cudaSupport (
     with cudaPackages;
@@ -71,7 +74,9 @@ buildPythonPackage {
     ]
   );
 
-  nativeBuildInputs = [ which ];
+  nativeBuildInputs = [ which ] ++ lib.optionals cudaSupport (with cudaPackages; [
+    cuda_nvcc
+  ]);
 
   propagatedBuildInputs = [
     numpy

